plugins/x86/x86_backend.ml

```diff
@@ -1,50 +1,61 @@
 open Core_kernel.Std
+open Or_error
 open Bap.Std
 open X86_tools_types
 
 module Insn = Disasm_expert.Basic.Insn
 module Table = String.Table
 
-module type Data = sig
-  val table : lifter Table.t
+type opcode = string
+type prefix = X86_prefix.t
+type pref_lift = mem -> bil -> bil
+
+module type S = sig
+  val register : opcode -> lifter -> unit
+  val handle_prefix : opcode -> prefix -> pref_lift -> unit    
+  module Make (T : Target) : Target
 end
 
-module Make (D : Data) (PR : PR) = struct
-  open D
-  let register op lift =
-    let lift mem insn =
-      try
-        let open Or_error in
-        match PR.parse mem op with
-        | `LOCK_PREFIX -> lift mem insn >>| PR.lock
-        | `REP_PREFIX ->  lift mem insn >>| PR.rep mem
-        | `REPE_PREFIX ->  lift mem insn >>| PR.repe mem
-        | `REPNE_PREFIX -> lift mem insn >>| PR.repne mem
-        | `EXCEPTION no -> Ok [ Bil.cpuexn no ]
-        | `NONE -> lift mem insn
-      with exn -> Or_error.of_exn exn in
-    String.Table.set
-      table
-      ~key:(X86_opcode.sexp_of_t op |> Sexp.to_string)
-      ~data:lift
+module Make (PR : PR) = struct
+
+  type record = {
```

Comment:
I would say that approach to register prefix handlers is wrong. because at first we register just one more function with our lifter function - so a question why a lifter can't do a whole work and treat prefixes as needed? and at second, handler `prefix -> bil -> bil` will not suitable for all types of prefixes, e.g. address or data size overriding prefixes, since instruction `bil` code will be dependent from prefix, i.e. will be wrong to generate `bil` code for instruction and prefix separately.